lib/src/subscription.dart

```diff
@@ -0,0 +1,216 @@
+  /// Finds a subscription by query.
+  ///
+  /// If the Subscription set does not contain a subscription with the provided query,
+  /// return null
+  Subscription? find<T extends RealmObject>(RealmResults<T> query) {
+    return realmCore.findSubscriptionByQuery(this, query).convert(Subscription._);
+  }
+
+  /// Finds an existing [Subscription] in this set by name.
+  ///
+  /// If the Subscription set does not contain a subscription with the provided name,
+  /// return null
+  Subscription? findByName(String name) {
+    return realmCore.findSubscriptionByName(this, name).convert(Subscription._);
+  }
+
+  Future<SubscriptionSetState> _waitForStateChange(SubscriptionSetState state) async {
+    final result = await realmCore.waitForSubscriptionSetStateChange(this, state);
+    realmCore.refreshSubscriptionSet(this);
+    return result;
+  }
+
+  /// Waits for the server to acknowledge the subscription set and return the matching objects.
+  ///
+  /// If the [state] of the subscription set is [SubscriptionSetState.complete]
+  /// the returned [Future] will complete immediately. If the state is
+  /// [SubscriptionSetState.error], the returned future will be throw an
+  /// error.
+  Future<SubscriptionSetState> waitForSynchronization() => _waitForStateChange(SubscriptionSetState.complete);
+
+  @override
+  int get length => realmCore.getSubscriptionSetSize(this);
+
+  @override
+  Subscription elementAt(int index) {
+    return Subscription._(realmCore.subscriptionAt(this, index));
+  }
+
+  /// Gets the [Subscription] at the specified index in the set.
+  Subscription operator [](int index) => elementAt(index);
+
+  @override
+  _SubscriptionIterator get iterator => _SubscriptionIterator._(this);
+
+  /// Update the subscription set and send the request to the server in the background.
+  ///
+  /// Calling [update] is a prerequisite for mutating the subscription set,
+  /// using a [MutableSubscriptionSet] parsed to [action].
+  ///
+  /// If you want to wait for the server to acknowledge and send back the data that matches the updated
+  /// subscriptions, use [waitForSynchronization].
+  void update(void Function(MutableSubscriptionSet mutableSubscriptions) action);
+
+  /// Gets the version of the subscription set.
+  int get version => realmCore.subscriptionSetGetVersion(this);
+
+  /// Gets the state of the subscription set.
+  SubscriptionSetState get state => realmCore.subscriptionSetGetState(this);
+}
+
+extension SubscriptionSetInternal on SubscriptionSet {
+  Realm get realm => _realm;
+  SubscriptionSetHandle get handle => _handle;
+
+  static SubscriptionSet create(Realm realm, SubscriptionSetHandle handle) => _ImmutableSubscriptionSet._(realm, handle);
+}
+
+class _ImmutableSubscriptionSet extends SubscriptionSet {
+  _ImmutableSubscriptionSet._(Realm realm, SubscriptionSetHandle handle) : super._(realm, handle);
+
+  @override
+  void update(void Function(MutableSubscriptionSet mutableSubscriptions) action) {
```

Comment:
I'm a little bit confused. We have ImmutableSubscriptionSet that allows to be updated and uses internally MutableSubscriptionSet. What is the purpose?

Reply:
You give a callback to update that is parsed a `MutableSubscriptionSet`. Only these have `add`/`remove*` methods. So you need to call `update` to actually obtain a mutable set, and you need to parse the `action` to actually do the update.

Reply:
OK